Badge + docs updates (#348)
drivendata/cookiecutter-data-science: docs/docs/using-the-template.md

# Using the template

> Ok, so you've created your project. You've [read the opinions section](opinions.md). You're ready to start doing some work.

Reviewer: ```suggestion
You've [created](index.md#starting-a-new-project) your project. You've [read the opinions section](opinions.md). You're ready to start doing some work.
```

> Here's a quick guide of the kinds of things we do once our project is ready to go.

Reviewer: ```suggestion
Here's a quick guide to the kinds of things we do once our project is ready to go.
```

## Set up version control

> Once you `cd` into the directory created by `ccds` you're ready to go.
> 

Reviewer: ```suggestion
```

Often, we start by initializing a `git` repository to track the code we write in version control and collaborate with teammates. At the command line, you can do this with the following commands which do the following: turn the folder into a git repository, add all of the files and folders created by CCDS into source control (except for what is in the `.gitignore` file), and then make a commit to the repository.

> ```bash
> git init

Reviewer: ```suggestion
```bash
# From inside your newly created project directory
git init
```

git add .
git commit -m "CCDS defaults"
```

> We usually commit the entire default CCDS structure so it is easy to track in verison history the changes we make to the structure separately.

Reviewer: Exact meaning felt a little unclear here, I attempted a rephrase.
```suggestion
We usually commit the entire default CCDS structure so it is easy to track the changes we make to the structure in version history.
```

> Now that default layout is committed, you should push it to a shared repository. You can do this through the interface of whatever source control platform you use. This may be GitHub, GitLab, Bitbucket, or something else.

Reviewer: ```suggestion
Now that the default layout is committed, you should push it to a shared repository. You can do this through the interface of whatever source control platform you use. This may be GitHub, GitLab, Bitbucket, or something else.
```

```bash
gh repo create
```

You'll be asked a series of questions to set up the repository on GitHub, for example:

```
gh repo create
? What would you like to do? Push an existing local repository to GitHub
? Path to local repository: .  # CCDS generated folder you are in
? Repository name: testccds    # Name of the project 
? Repository owner: pjbull     # Github username
? Description: Testing         # A short description of the project
? Visibility: Private          # 
✓ Created repository pjbull/testccds on GitHub
? Add a remote? Yes
? What should the new remote be called? origin
\✓ Added remote https://github.com/pjbull/testccds.git
? Would you like to push commits from the current branch to "origin"? Yes
Enumerating objects: 23, done.
Counting objects: 100% (23/23), done.
Delta compression using up to 12 threads
Compressing objects: 100% (20/20), done.
Writing objects: 100% (23/23), 6.47 KiB | 6.47 MiB/s, done.
Total 23 (delta 0), reused 0 (delta 0), pack-reused 0
To https://github.com/pjbull/testccds.git
 * [new branch]      HEAD -> main
branch 'main' set up to track 'origin/main'.
✓ Pushed commits to https://github.com/pjbull/testccds.git

> ```

Reviewer: I feel as though this is a little too much to provide here. It might be better if we just linked out to the gh cli docs.

## Create a Python virutal environment

We often use Python for our data science projects. We use a virtual environment to manage the packages we use in our project. This is a way to keep the packages we use in our project separate from the packages we use in other projects. This is especially important when we are working on multiple projects at the same time.

> Cookicutter Data Science supports a few options for virtual environment management, but no matter which you choose, you can create an environment with the following commands:

Reviewer: ```suggestion
Cookicutter Data Science supports [a few options](opinions.md#build-from-the-environment-up) for Python virtual environment management, but no matter which you choose, you can create an environment with the following commands:
```

```bash
make create_environment
```

Once the environment is created, you'll want to make sure to activate it. You'll have to do this following the instructions for your specific environment manager. We recommend using a shell prompt that shows you which environment you are in, so you can easily tell if you are in the right environment, for example [starship](https://starship.rs/). You can also use the command `which python` to make sure that your shell is pointing to the version of Python associated with your virtual environment.

Once you are sure that your environment is activated in your shell, you can install the packages you need for your project. You can do this with the following command:

```bash
make requirements
```

## Add your data

There's no universal advice for how to manage your data, but here are some recommendations for starting points depending on where the data comes from:

 - **Flat files (e.g., CSVs or spreadsheets) that are static** - Put these files into your `data/raw` folder and then run `make sync_data_up` to push the raw data to your cloud provider.

>  - **Flat files that change and are extracted from somewhere** - Add a Python script to your source module in `data/make_dataset.py` that downloads the data and puts it in the `data/raw` folder. Then you can use this to get the latest and push it up to your cloud host as it changes.

Reviewer: I know we say folks should read the opinions section first, but I feel like it can't hurt to link out to the relevant sections / give people reminders to be a little careful
```suggestion
 - **Flat files that change and are extracted from somewhere** - Add a Python script to your source module in `data/make_dataset.py` that downloads the data and puts it in the `data/raw` folder. Then you can use this to get the latest and push it up to your cloud host as it changes (be careful not to [override your raw data](opinions.md/#data-analysis-is-a-directed-acyclic-graph)).
```

 - **Databases you connect to with credentials** - Store your credentials in `.env`. We recommend adding a `db.py` file or similar to your `data` module that connects to the database and pulls data. If your queries generally fit into memory, you can just have functions in the `db.py` to load data that you use in analysis. If not, you'll want to add a script like above to download the data to the `data/raw` folder.

## Check out a branch

> We'll talk about code review later, but it's a good practice to use feature branches and pull requests to keep your development organized. Now that you have source control configured, you can check out a branch to work in with:

Reviewer: ```suggestion
We'll talk about code review later, but it's a good practice to use feature branches and pull requests to keep your development organized. Now that you have source control configured, you can check out a branch to work with:
```

```
git checkout -b initial-exploration
```

## Open a notebook

> 

Reviewer: We're mostly agnostic about notebooks but this section uses Jupyter, may be worthwhile to have something like

```suggestion

!!! note 

    The following assumes you're using a Jupyter notebook, but while the specific commands for another notebook tool may look a little bit different, the process guidance still applies.

```

    - `1` - Data cleaning and feature creation - often writes data to `data/processed` or `data/interim`
    - `2` - Visualizations - often writes publication-ready viz to `reports`
    - `3` - Modeling - training machine learning models
    - `4` - Publication - Notebooks that get turned directly into reports
- `pjb` - Your initials; this is helpful for knowing who created the notebook and prevents collisions from people working in the same notebook.
- `data-source-1` - A description of what the notebook covers 

Now that you have your notebook going, start your analysis!

## Refactoring code into shared modules

As your project goes on, you'll want to refactor your code in a way that makes it easy to share between notebooks and scripts. We recommend creating a module in the `{{ cookiecutter.module_name }}` folder that contains the code you use in your project. This is a good way to make sure that you can use the same code in multiple places without having to copy and paste it.

Because the default structure is a Python package and is installed by default, you can do the following to make that code available to you within a Jupyter notebook.

> First, we recommend turning on the `autoreload` extension. This will make Jupyter always go back to the source code for the module rather than caching it in memory. If your notebook isn't reflecting the latest changes from your changes to a `.py` file, try restarting the kernel and make sure `autoreload is on. We add a cell at the top of the notebook with the following:

Reviewer: ```suggestion
First, we recommend turning on the `autoreload` extension. This will make Jupyter always go back to the source code for the module rather than caching it in memory. If your notebook isn't reflecting the latest changes from your changes to a `.py` file, try restarting the kernel and make sure `autoreload` is on. We add a cell at the top of the notebook with the following:
```

```
%load_ext autoreload
%autoreload 2
```

> Now all your code should be importable. At the start of the CCDS project, you picked a module name. It's the same name as folder that is in the root project directory. For example, if the module name were `my_project` you could use code by importing it like:

Reviewer: ```suggestion
Now all your code should be importable. At the start of the CCDS project, you picked a module name. It's the same name as the folder that is in the root project directory. For example, if the module name were `my_project` you could use code by importing it like:
```
